kml/serdes_gpx.cpp

```diff
@@ -282,43 +291,64 @@ void GpxParser::CharData(std::string value)
     std::string_view const & currTag = m_tags[count - 1];
     std::string_view const & prevTag = m_tags[count - 2];
 
-    if (prevTag == gpx::kWpt)
-    {
-      if (currTag == gpx::kName)
-        m_name[kml::kDefaultLang] = value;
-      else if (currTag == gpx::kDesc)
-        m_description[kml::kDefaultLang] = value;
-    }
-    else if (prevTag == gpx::kTrk || prevTag == gpx::kRte)
-    {
-      if (currTag == gpx::kName)
-      {
-        m_name[kml::kDefaultLang] = value;
-        if (m_categoryData->m_name[kml::kDefaultLang].empty())
-          m_categoryData->m_name[kml::kDefaultLang] = value;
-      }
-      else if (currTag == gpx::kDesc)
-      {
-        m_description[kml::kDefaultLang] = value;
-        if (m_categoryData->m_description[kml::kDefaultLang].empty())
-          m_categoryData->m_description[kml::kDefaultLang] = value;
-      }
-    }
-    else if (prevTag == gpx::kMetadata)
-    {
-      if (currTag == gpx::kName)
-        m_categoryData->m_name[kml::kDefaultLang] = value;
-      else if (currTag == gpx::kDesc)
-        m_categoryData->m_description[kml::kDefaultLang] = value;
-    }
-    if (currTag == gpx::kGarminColor)
+    if (currTag == gpx::kName)
+      ParseName(value, prevTag);
+    else if (currTag == gpx::kDesc)
+      ParseDescription(value, prevTag);
+    else if (currTag == gpx::kGarminColor)
       ParseGarminColor(value);
     else if (currTag == gpx::kOsmandColor)
       ParseOsmandColor(value);
     else if (currTag == gpx::kColor)
       ParseColor(value);
+    else if (currTag == gpx::kEle)
+      ParseAltitude(value);
   }
 }
+
+void GpxParser::ParseDescription(std::string const & value, std::string_view prevTag)
+{
+  if (prevTag == kWpt)
+  {
+    m_description[kDefaultLang] = value;
+  }
+  else if (prevTag == kTrk || prevTag == kRte)
+  {
+    m_description[kDefaultLang] = value;
+    if (m_categoryData->m_description[kDefaultLang].empty())
+      m_categoryData->m_description[kDefaultLang] = value;
+  }
+  else if (prevTag == kMetadata)
+  {
+    m_categoryData->m_description[kDefaultLang] = value;
+  }
+}
+
+void GpxParser::ParseName(std::string const & value, std::string_view prevTag)
+{
+  if (prevTag == kWpt)
+  {
+    m_name[kDefaultLang] = value;
+  }
+  else if (prevTag == kTrk || prevTag == kRte)
+  {
+    m_name[kDefaultLang] = value;
+    if (m_categoryData->m_name[kDefaultLang].empty())
+      m_categoryData->m_name[kDefaultLang] = value;
+  }
+  else if (prevTag == kMetadata)
+  {
+    m_categoryData->m_name[kDefaultLang] = value;
+  }
+}
+
+void GpxParser::ParseAltitude(std::string const & value)
+{
+  double rawAltitude;
+  if (strings::to_double(value, rawAltitude))
+    m_altitude = static_cast<geometry::Altitude>(round(rawAltitude));
```

Comment:
Could a point without an altitude get an old altitude from a previous point with altitude?

Reply:
Definitely should add else m_altitude  = kInvalidAltitude

Reply:
Fixed and added UNIT_TEST(Gpx_Altitude_Issues)